internal/peermanager/peerwantmanager.go

```diff
@@ -79,26 +90,30 @@ func (pwm *peerWantManager) removePeer(p peer.ID) {
 // PrepareBroadcastWantHaves filters the list of want-haves for each peer,
 // returning a map of peers to the want-haves they have not yet been sent.
 func (pwm *peerWantManager) prepareBroadcastWantHaves(wantHaves []cid.Cid) map[peer.ID][]cid.Cid {
-	res := make(map[peer.ID][]cid.Cid)
+	res := make(map[peer.ID][]cid.Cid, len(pwm.peerWants))
+	for _, c := range wantHaves {
+		if pwm.broadcastWants.Has(c) {
+			// Already a broadcast want, skip it.
+			continue
+		}
+		pwm.broadcastWants.Add(c)
+
+		// Prepare broadcast.
+		wantedBy := pwm.wantPeers[c]
+		for p := range pwm.peerWants {
+			// If we've already sent a want to this peer, skip them.
+			//
+			// This is faster than checking the actual wantlists due
+			// to better locality.
+			if _, ok := wantedBy[p]; ok {
+				continue
+			}
 
-	// Iterate over all known peers
-	for p, pws := range pwm.peerWants {
-		// Iterate over all want-haves
-		for _, c := range wantHaves {
-			// If the CID has not been sent as a want-block or want-have
-			if !pws.wantBlocks.Has(c) && !pws.wantHaves.Has(c) {
-				// Record that the CID has been sent as a want-have
-				pws.wantHaves.Add(c)
-
-				// Update the reverse index
-				pwm.reverseIndexAdd(c, p)
-
-				// Add the CID to the results
-				if _, ok := res[p]; !ok {
-					res[p] = make([]cid.Cid, 0, 1)
-				}
-				res[p] = append(res[p], c)
+			cids, ok := res[p]
+			if !ok {
+				cids = make([]cid.Cid, 0, len(wantHaves))
```

Comment:
Pre-allocate assuming we're going to send this peer all the wants (common case).